client reset step1 - cleanup

ironage opened the pull request:
This is a first step towards https://jira.mongodb.org/browse/RCORE-679

- Remove `ClientResyncMode::Recover` option. This doesn't work and shouldn't be used by anyone currently.
- In a client reset, remove the CLIENT_VERSION request; it isn't used.
- Remove the STATE message request and handling from the client and sever. The current server was always responding with empty state. Clients will now consider receiving this message as an error because they never request it.
- Handling for async open, involving state files is mostly removed.
- When a client reset happened, we would create a metadata Realm to handle state download but this isn't needed anymore so it is removed.

jbreams commented on src/realm/sync/noinst/client_reset_operation.cpp:
```diff
@@ -0,0 +1,61 @@
+#include <realm/sync/noinst/client_reset.hpp>
+#include <realm/sync/noinst/client_reset_operation.hpp>
+
+using namespace realm;
```
Instead of doing `using namespace` for the bodies of implementation files, let's just enclose the implementation in the correct namespace. You can do
```
namespace realm::_impl {
body of cpp file
} // namespace realm::_impl
```

That way we don't risk adding any globals to the top-level global scope.

jbreams commented on src/realm/sync/noinst/client_reset_operation.hpp:
```diff
@@ -0,0 +1,73 @@
+#include <realm/sync/protocol.hpp>
+#include <realm/util/aes_cryptor.hpp>
+
+namespace realm {
```
You can do `namespace realm::_impl` here.

jbreams reviewed (changes requested):
First round of comments here. Looks like we're almost there!

ironage commented:
@jbreams, I have addressed your change requests.

ironage commented:
The additional client reset tests that I added back in revealed a segfault in MSVC 2019 Release builds. The fix that I provided in 28f1ab9 is a workaround for the compiler's behaviour. I suspect the change works because MSVC might be implicitly converting the `int64_t` handling into `GlobalKey(Mixed(int64_t))` and only providing a single GlobalKey case for both int64_t and GlobalKey while `mpark::visit` generates code for both explicitly. I haven't spent the time to try to analyze the compiler output so that's just my best guess. But I can confirm that the change fixes the segfault in Release mode and the stack smashing exception in Debug mode by testing the changes on a spawned host.  It looks like this should fix https://github.com/realm/realm-core/issues/4624.

ironage commented:
I have extracted the bug fix into a separate PR https://github.com/realm/realm-core/pull/4825 and reverted the fix from these changes to keep things clean. This means that we will delay this PR until #4825 is merged or resolved in some way because the tests on this branch will fail until the issue is fixed.
